File: merge_sort.c
/*Copyright to 20145523 KimSangHeon
Last updated date : 2018-03-17*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

//Print Result Function
void PrintArr ( int* input, int N ) {
    int i;
    for ( i = 0; i < N; i++) {
        printf("%d\n", input[i]);
    }
}
//Merge Function
void Merge ( int* input, int low, int mid, int high ) {
    int i, j, k;
    int *tmp;

    tmp = (int*)malloc( (high - low + 1) * sizeof(int));
    i = low;
    j = mid + 1;
    k = 0;
    //Compare Left block and Right block
    while ( i <= mid && j <= high ) {
        if ( input[i] > input[j] ) {
            tmp[k++] = input[i++];
        }
        else {
            tmp[k++] = input[j++];
        }
    }
    //If still remaining instances are exist
    while (i <= mid) tmp[k++] = input[i++];
    while (j <= high) tmp[k++] = input[j++];
    
    for ( i = 0; i < k; i++) {
        input[low + i] = tmp[i];
    }

    free(tmp);
}
//Merge Sorting Function
void MergeSort ( int* input, int low, int high ) {
    int mid;
    if(high > low) {
        mid = (low + high) / 2;
        // Keep spliting
        MergeSort(input, low, mid);
        MergeSort(input, mid + 1, high);
        // 
        Merge(input, low, mid, high);
    }
}

int main ( int argc, char* argv[] ) {
    // ERROR : Wrong command line input
    if (argc != 3 ) exit(EXIT_FAILURE);
    clock_t startTime;
    int i, N, input, count;
    int* inputN, *inputORG;
    double endTime;
    char* ptr;
    FILE *fp;

    N = strtol(argv[1], &ptr, 10);
    count = 0;
    if ( *ptr != '\0' ) exit(EXIT_FAILURE); // ERROR : If argv[1] is not integer
    if ( N == 0 ) exit(EXIT_FAILURE); // ERROR : If argv[1] is 0 
    inputN = (int*)malloc(sizeof(int) * N);
    inputORG = (int*)malloc(sizeof(int) * N);

    fp = fopen(argv[2], "r");
    // ERROR : If file doesn't exist
    if ( fp == NULL ) exit(EXIT_FAILURE);
    for ( i = 0; i < N; i++ ) {
        if ((fscanf(fp, "%d\n", &input)) == EOF ){//If N>K
            N = i;
            break;
        }
        inputN[i] = input;
    }
    fclose(fp);
    //Save original data to another inputORG
    memcpy(inputORG, inputN, N * sizeof(int));

    startTime = clock(); 
    MergeSort(inputN, 0, N - 1);
    // ERROR : If elapsed time is less then 1000ms
    if ( (clock() - startTime) < 1000 ) {
        do {
            count++;
            // Copy origin data to inputN whenever loop started
            memcpy(inputN, inputORG, N * sizeof(int)); 
            MergeSort(inputN, 0, N - 1);
        } while ( clock() - startTime < 1000 );
    }
    else count++;
    endTime = (double)(clock() - startTime) / (CLOCKS_PER_SEC / 1000);

    PrintArr(inputN, N);
    printf("Running time = %f ms\n", (double)(endTime / count) );

    free(inputN);
    free(inputORG);
    return 0;
}

File: merge_sort.py
'''
20145523 KimSangHeon
Last updated date : 2018-03-09
'''
import sys
import time
import gc

#Print Result Function
def printArr(inputN):
    for i in inputN:
        print(i)
#Merge Sorting Function
def MergeSort(inputN, N):
    if N < 2: return
    mid = int(N / 2)
    #Split input Array into Two arrays
    Left = inputN[:mid]
    Right = inputN[mid:]
    #Keep Spliting till less then two components left
    MergeSort(Left, mid)
    MergeSort(Right, N - mid)
    #Merge splited arrays
    Merge(inputN, Left, Right, N)

def Merge(inputN, Left, Right, N):
    i = 0
    j = 0
    k = 0
    LeftN = int(N / 2)
    RightN = N - LeftN
    while i < LeftN and j < RightN:
        if Left[i] > Right[j]:
            inputN[k] = Left[i]
            k = k + 1
            i = i + 1
        else:
            inputN[k] = Right[j]
            k = k + 1
            j = j + 1
    #If still there is remaining things are exist
    while i < LeftN : 
        inputN[k] = Left[i]
        k = k + 1
        i = i + 1
    while j < RightN :
        inputN[k] = Right[j]
        k = k + 1
        j = j + 1

#Main part
if len(sys.argv) != 3: sys.exit('wrong input length') #ERROR : Wrong command line input
try :
    fp = open(sys.argv[2],"r")
except FileNotFoundError as e:
    print(str(e)) # ERROR : If file doesn't exist
try:
    N = int(sys.argv[1])
except ValueError as e:
    print(str(e)) # ERROR : If argv[1] is not integer
if N == 0 : sys.exit(0) # ERROR : If argv[1] is 0
inputN = []

#read data
for i in range(0, N):
    try:
        inp = int(fp.readline())
        if inp == '':#if N > K
            N = i
            break
    # ERROR : If file has data which is not integer
    except ValueError as e:
        print(str(e))

    inputN.append(inp)
inputTMP = inputN[:]
fp.close()
count = 0

startTime = time.monotonic()
MergeSort(inputN, N)
# ERROR : If elapsed time is less then 1000ms
if (time.monotonic() - startTime) < 1.000:
    while time.monotonic() - startTime < 1.000:
        count += 1
        # Copy origin data to inputN whenever loop started
        inputN = inputTMP[:]
        MergeSort(inputN, N)
else:
    count = 1
endTime = time.monotonic() - startTime

printArr(inputN)
print("Running time = " + str(endTime * 1000 / count) + " ms")
gc.collect()# free any kind of garbage memory